core/state/state_object.go

```diff
@@ -57,118 +57,166 @@ func (self Storage) Copy() Storage {
+// Finally, call CommitTrie to write the modified storage trie into a database.
 type StateObject struct {
-	trie *trie.SecureTrie
-
-	// Address belonging to this account
-	address common.Address
-	// The balance of the account
-	balance *big.Int
-	// The nonce of the account
-	nonce uint64
-	// The code hash if code is present (i.e. a contract)
-	codeHash []byte
-	// The code for this account
-	code Code
-	// Cached storage (flushed when updated)
-	storage Storage
-
-	// Mark for deletion
+	address common.Address // Ethereum address of this account
+	data    accountData
+
+	// DB error.
+	// State objects are used by the consensus core and VM which are
+	// unable to deal with database-level errors. Any error that occurs
+	// during a database read is memoized here and will eventually be returned
+	// by StateDB.Commit.
+	dbErr error
+
+	// Write caches.
+	trie    *trie.SecureTrie // storage trie, which becomes non-nil on first access
+	code    Code             // contract bytecode, which gets set when code is loaded
+	storage Storage          // Cached storage (flushed when updated)
+
+	// Cache flags.
 	// When an object is marked for deletion it will be delete from the trie
 	// during the "update" phase of the state transition
+	dirty   bool // true if anything has changed
 	remove  bool
 	deleted bool
-	dirty   bool
+}
+
+// accountData is the Ethereum consensus representation of accounts.
+// These objects are stored in the main account trie.
+type accountData struct {
+	Nonce    uint64
+	Balance  *big.Int
+	Root     common.Hash // merkle root of the storage trie
+	CodeHash []byte
 }
 
 func NewStateObject(address common.Address, db trie.Database) *StateObject {
 	object := &StateObject{
-		address:  address,
-		balance:  new(big.Int),
-		dirty:    true,
-		codeHash: emptyCodeHash,
-		storage:  make(Storage),
+		address: address,
+		data:    accountData{Balance: new(big.Int), CodeHash: emptyCodeHash},
+		storage: make(Storage),
+		dirty:   true,
 	}
-	object.trie, _ = trie.NewSecure(common.Hash{}, db)
 	return object
 }
 
+// setError remembers the first non-nil error it is called with.
+func (self *StateObject) setError(err error) {
+	if self.dbErr == nil {
+		self.dbErr = err
+	}
+}
+
 func (self *StateObject) MarkForDeletion() {
 	self.remove = true
 	self.dirty = true
 
 	if glog.V(logger.Core) {
-		glog.Infof("%x: #%d %v X\n", self.Address(), self.nonce, self.balance)
+		glog.Infof("%x: #%d %v X\n", self.Address(), self.Nonce(), self.Balance())
 	}
 }
 
-func (c *StateObject) getAddr(addr common.Hash) common.Hash {
-	var ret []byte
-	rlp.DecodeBytes(c.trie.Get(addr[:]), &ret)
-	return common.BytesToHash(ret)
-}
-
-func (c *StateObject) setAddr(addr, value common.Hash) {
-	v, err := rlp.EncodeToBytes(bytes.TrimLeft(value[:], "\x00"))
-	if err != nil {
-		// if RLPing failed we better panic and not fail silently. This would be considered a consensus issue
-		panic(err)
+func (c *StateObject) initTrie(db trie.Database) {
```

Comment:
This should really , really be an accessor that does the lazy initialization, or we're going to end up accessing the trie somewhere without remembering to initialize it. If there's not time to fix, can we leave a comment somewhere to fix it?